components/notebook-controller/config/overlays/openshift/generate-metadata-yaml.sh

```diff
@@ -0,0 +1,165 @@
+#!/usr/bin/env bash
+
+set -uo pipefail
+
+function trap_exit() {
+	rc=$?
+
+	set +x
+
+	exit $rc
+}
+
+trap "trap_exit" EXIT
+
+_derive_metadata()
+{
+  # inspired from https://stackoverflow.com/a/29835459
+  current_dir=$(CDPATH= cd -- "$(dirname -- "$0")" && pwd)
+
+  kf_project_file="${current_dir}/PROJECT"
+  if [ -e "${kf_project_file}" ]; then
+
+    if [ -z "${metadata_repo_url}" ]; then
+      project_repo_reference=$(yq -e '.repo' "${kf_project_file}")
+      project_repo_parts=( $(printf "%s" ${project_repo_reference##https://} | tr '/' ' ') )
+      github_host="${project_repo_parts[0]}"
+      github_owner="${project_repo_parts[1]}"
+      github_repo="${project_repo_parts[2]}"
+
+      metadata_repo_url=$(printf "https://%s/%s/%s" "${github_host}" "${github_owner}" "${github_repo}")
+    fi
+
+    if [ -z "${metadata_name}" ]; then
+      project_domain=$(yq -e '.domain' "${kf_project_file}")
+      project_name=$(yq -e '.projectName' "${kf_project_file}")
+      metadata_name="${project_domain} ${project_name}"
+    fi
+
+  fi
+
+  if [ -z "${metadata_version}" ]; then
+    repo_root=$(git rev-parse --show-toplevel)
+    version_file="${repo_root}/releasing/version/VERSION"
+
+    metadata_version=$(cat "${version_file}" | head -n 1)
+  fi
+}
+
+_fallback_to_existing_values()
+{
+  if [ -n "${existing_fallback}" ]; then
+    if [ -z "${metadata_repo_url}" ]; then
+      metadata_repo_url=$(yq -e '.releases[0].repoUrl' "${output_file}")
+    fi
+
+    if [ -z "${metadata_version}" ]; then
+      metadata_version=$(yq -e '.releases[0].version' "${output_file}")
+    fi
+
+    if [ -z "${metadata_name}" ]; then
+      metadata_name=$(yq -e '.releases[0].name' "${output_file}")
+    fi
+  fi
+}
+
+_check_for_missing_data()
+{
+  missing_data=
+
+  if [ -z "${metadata_repo_url}" ]; then
+    printf "%s\n" "repoUrl attribute not specified and unable to be inferred"
+    missing_data=1
+  fi
+
+  if [ -z "${metadata_version}" ]; then
+    printf "%s\n" "version attribute not specified and unable to be inferred"
+    missing_data=1
+  fi
+
+  if [ -z "${metadata_name}" ]; then
+    printf "%s\n" "name attribute not specified and unable to be inferred"
+    missing_data=1
+  fi
+
+  if [ -n "${missing_data}" ]; then
+    exit 1
+  fi
+}
+
+_handle_metadata_file()
+{
+
+  _derive_metadata
+
+  _fallback_to_existing_values
+
+  _check_for_missing_data
+
+  # NOTE: Does not handle multiple entries!!
+  yq_env_arg="${metadata_name}" yq -i '.releases[0].name = strenv(yq_env_arg)' "${output_file}"
+  yq_env_arg="${metadata_version}" yq -i '.releases[0].version = strenv(yq_env_arg)' "${output_file}"
+  yq_env_arg="${metadata_repo_url}" yq -i '.releases[0].repoUrl = strenv(yq_env_arg)' "${output_file}"
```

Comment:
**`"Controversial" Decision #3`**

While the specification for `odh_metadata.yaml` supports multiple elements in the `releases` attribute - w/in the kubeflow `components` directory of `red-hat-data-services` - it will always map 1:1... so I simply hardcoded the script to always interact with the 0th element.

This simplifies logic greatly and I don't think in practice, for rhds kubeflow, we'd ever need anything more complicated.

Good idea? Bad Idea? _You tell me!_